apps/settings/js/modules/addon_manager.js

```diff
@@ -0,0 +1,201 @@
+ * installed addons, enabling, disabling, and uninstalling addons, and
+ * for obtaining a list of apps targeted by a specified add-on. It also
+ * emits events when the set of installed addons has changed and when
+ * the set of installed apps has changed (because this might affect the
+ * set of targeted apps for an addon).
+ *
+ * AddonManager uses the AppsCache module for efficient access to the full
+ * list of installed apps.
+ */
+/* global MozActivity */
+define(function(require) {
+  'use strict';
+
+  var AppsCache = require('modules/apps_cache');
+
+  // This is the AddonManager API
+  var AddonManager = {
+    getAddons: getAddons,               // returns promise<app[]>
+    isAddon: isAddon,                   // returns true or false for app
+    isEnabled: isEnabled,               // returns true or false for addon
+    enableAddon: enableAddon,           // pass an addon object to enable
+    disableAddon: disableAddon,         // pass an addon to disable
+    canDelete: canDelete,               // returns true or false for addon
+    deleteAddon: deleteAddon,           // pass an app, returns promise
+    shareAddon: shareAddon,             // returns a promise
+    getAddonTargets: getAddonTargets,   // pass an app, returns a promise<app[]>
+    addEventListener: addListener,      // For event type "addonschanged"
+    removeEventListener: removeListener
+  };
+
+  function getManifest(app) {
+    return app.manifest || app.updateManifest;
+  }
+
+  function isAddon(app) {
+    var manifest = getManifest(app);
+    return manifest && manifest.role === 'addon';
+  }
+
+  function getAddons() {
+    return AppsCache.apps().then(function(apps) {
+      return apps.filter(isAddon);
+    });
+  }
+
+  function isEnabled(addon) {
+    return isAddon(addon) && addon.enabled;
+  }
+
+  function enableAddon(addon) {
+    if (isAddon(addon)) {
+      navigator.mozApps.mgmt.setEnabled(addon, true);
+    }
+  }
+
+  function disableAddon(addon) {
+    if (isAddon(addon)) {
+      navigator.mozApps.mgmt.setEnabled(addon, false);
+    }
+  }
+
+  function canDelete(addon) {
+    return isAddon(addon) && addon.removable;
+  }
+
+  function deleteAddon(addon) {
+    if (!isAddon(addon)) {
+      return Promise.reject('not an addon');
+    }
+
+    if (!addon.removable) {
+      return Promise.reject('addon is not deletable');
+    }
+
+    return new Promise(function(resolve, reject) {
+      var request = navigator.mozApps.mgmt.uninstall(addon);
+      request.onsuccess = function() {
+        resolve();
+      };
+      request.onerror = function() {
+        reject();
+      };
+    });
+  }
+
+  function shareAddon(addon) {
+    if (!isAddon(addon)) {
+      return Promise.reject('not an addon');
+    }
+
+    return new Promise(function(resolve, reject) {
+      var activity = new MozActivity({
+        name: 'share',
+        data: {
+          type: 'app',
+          app: addon.origin
```

Comment:
you need to use the manifestURL, not the origin since we can have multiple apps per origin.